src/stores/external-source.ts

```diff
@@ -0,0 +1,170 @@
+import { derived, writable, type Writable } from 'svelte/store';
+import {
+  type DerivationGroup,
+  type ExternalSourcePkey,
+  type ExternalSourceSlim,
+  type ExternalSourceType,
+  type PlanDerivationGroup,
+  type UserSeenEntry,
+} from '../types/external-source';
+import gql from '../utilities/gql';
+import { planId } from './plan';
+import { gqlSubscribable } from './subscribable';
+
+/* Writeable. */
+export const parsingError: Writable<string | null> = writable(null);
+export const creatingExternalSource: Writable<boolean> = writable(false);
+export const createExternalSourceError: Writable<string | null> = writable(null);
+export const createExternalSourceTypeError: Writable<string | null> = writable(null);
+export const createDerivationGroupError: Writable<string | null> = writable(null);
+export const derivationGroupPlanLinkError: Writable<string | null> = writable(null);
+export const getExternalSourceMetadataError: Writable<string | null> = writable(null);
+
+/* Subscriptions. */
+export const externalSources = gqlSubscribable<ExternalSourceSlim[]>(
+  gql.SUB_EXTERNAL_SOURCES,
+  {},
+  [],
+  null,
+  transformExternalSources,
+);
+export const externalSourceTypes = gqlSubscribable<ExternalSourceType[]>(gql.SUB_EXTERNAL_SOURCE_TYPES, {}, [], null);
+export const derivationGroups = gqlSubscribable<DerivationGroup[]>(
+  gql.SUB_DERIVATION_GROUPS,
+  {},
+  [],
+  null,
+  transformDerivationGroups,
+);
+export const planDerivationGroupLinks = gqlSubscribable<PlanDerivationGroup[]>(
+  gql.SUB_PLAN_DERIVATION_GROUP,
+  {},
+  [],
+  null,
+);
+
+// this tracks each user's view of the sources. if something exists in externalSources that isn't in usersSeenSources, it's treated as unseen, and vice versa for deleted.
+export const usersSeenSources = gqlSubscribable<Record<string, UserSeenEntry[]>>(
+  gql.SUB_SEEN_SOURCES,
+  {},
+  {},
+  null,
+  transformUsersSeenSources,
+);
+
+/* Derived. */
+export const selectedPlanDerivationGroupNames = derived(
+  [planDerivationGroupLinks, planId],
+  ([$planDerivationGroupLinks, $planId]) =>
+    $planDerivationGroupLinks.filter(link => link.plan_id === $planId).map(link => link.derivation_group_name),
+);
+export const selectedPlanDerivationGroupEventTypes = derived(
+  [derivationGroups, selectedPlanDerivationGroupNames],
+  ([$derivationGroups, $selectedPlanDerivationGroupIds]) => {
+    return $derivationGroups
+      .filter(dg => $selectedPlanDerivationGroupIds.includes(dg.name))
+      .map(dg => dg.event_types)
+      .reduce((acc, curr) => acc.concat(curr), []);
+  },
+);
+
+/* Helper Functions. */
+export function resetExternalSourceStores() {
+  createExternalSourceError.set(null);
+  createExternalSourceTypeError.set(null);
+  createDerivationGroupError.set(null);
+  derivationGroupPlanLinkError.set(null);
+}
+
+export function getRowIdFromExternalSourceId(externalSourceId: ExternalSourcePkey): string {
```

Comment:
This doesn't look like it's being used anywhere outside of this file, so this doesn't need to be exported. If you think that you'll need to use it outside of the file, then please move it to a util file instead.

Reply:
Yes!